add support for extensible enums

hugowetterberg opened the pull request:
Adds the option to specify stand-alone extensible enums: 

```json
  "enums": [
    {
      "declare": "core/text-roles",
      "name": "Text block roles",
      "values": {
        "heading-1": {},
        "heading-2": {},
        "heading-3": {},
        "heading-4": {},
        "preamble": {},
        "blockquote": {}
      }
    }
  ],
```

It's possible to add, deprecate or forbid values:

```json
  "enums": [
    {
      "match": "core/text-roles",
      "values": {
        "heading-3": {"forbidden":true},
        "heading-4": {"forbidden":true},
        "vignette": {},
        "dateline": {
          "deprecated": {
            "label": "dateline",
            "doc": "Use 'vignette' instead"
          }
        }
      }
    }
  ],
```

This is used to make it possible to create a core/text block and then extend it with different roles:

```json
        {
          "declares": {"type":"core/text"},
          "attributes": {
            "role": {
              "optional":true,
              "enumReference": "core/text-roles"
            }
          },
          "data": {
            "text":{
              "allowEmpty":true,
              "format": "html"
            }
          }
        },
```

A lot of the code changes are in service of making it possible for string constraints to return deprecation information.

As far as I could tell `DocumentConstraint.checkAttributes()` duplicated `validateDocumentAttributes()` so I removed that method and the call to it.

`checkDeprecation()` was extended to accept multiple deprecations.

Document and block attributes, which always are "present" given that they map to struct fields and not map entries like data attributes, are not collected by revisor as values any more. Empty data attributes are still treated as signal.

fred-o reviewed (approved):
Looks good to me. The only thing I could find was a missing comma on line 64 (between 'type' and 'rel')

hugowetterberg committed:
add support for extensible enums

hugowetterberg committed:
update dependencies

hugowetterberg committed:
having a nil deprecation is not an invalid value

(Deprecation, error) looks like a result-error pair to the linter, when it's
actually two classes of errors, one soft and one hard.

hugowetterberg committed:
update jsonschema
